SLS-712 Disable RTS in a disaggregated storage context

korteland opened the pull request.

agorrod commented on src/rollback_to_stable/rts_api.c:
```diff
@@ -115,6 +115,10 @@ __rollback_to_stable_int(WT_SESSION_IMPL *session, bool no_ckpt)
     WT_ASSERT_SPINLOCK_OWNED(session, &conn->schema_lock);
 
+    /* Check this nice and early, before we do any prep work. */
+    if (__wt_conn_is_disagg(session))
```
Could you add a log message here, just so that it gets flagged (probably at a higher than default log level)

korteland replied:
Done - see https://github.com/wiredtiger/wiredtiger/pull/11413#issuecomment-2588896404.

agorrod reviewed (changes requested):
I've got one tiny request, otherwise LGTM

quchenhao commented on src/rollback_to_stable/rts_api.c:
```diff
@@ -115,6 +115,10 @@ __rollback_to_stable_int(WT_SESSION_IMPL *session, bool no_ckpt)
     WT_ASSERT_SPINLOCK_OWNED(session, &conn->checkpoint_lock);
     WT_ASSERT_SPINLOCK_OWNED(session, &conn->schema_lock);
 
+    /* Check this nice and early, before we do any prep work. */
```
Should the check be here? This disables rts for all the situations. I thought we only want to disable it for recovery case. I think it is possible for a caller to call session->rts in the disaggregated storage. It will rollback any unstable updates on the update chain.

korteland replied:
Done - see https://github.com/wiredtiger/wiredtiger/pull/11413#issuecomment-2588896404.

korteland commented:
I've reworked this a bit. Chenhao and I agreed that we weren't confident enough yet about the runtime RTS situation, so I'm only disabling recovery RTS now.

The implementation isn't my favourite - we really ought to take a high-level review of the disagg config after the prototyping phase. The problem is that knowing we're disagg can only happen after we've configured disagg, which needs the metadata file. But having the metadata file relies on running recovery.

In addition, I've added some verbose messages when skipping RTS, and made the new test more robust.

quchenhao commented on src/txn/txn.c:
```diff
@@ -2772,7 +2772,7 @@ __wt_txn_global_shutdown(WT_SESSION_IMPL *session, const char **cfg)
          * clean shutdown.
          */
-        if (use_timestamp) {
+        if (use_timestamp && !__wt_conn_is_disagg(session)) {
```
Can we use a local variable like:

is_disagg = __wt_conn_is_disagg instead of calling it twice?

korteland replied:
Done in 3b0fb82.

quchenhao commented on test/suite/test_rollback_to_stable44.py:
```diff
@@ -0,0 +1,87 @@
+        self.session.commit_transaction('commit_timestamp=30')
+        c.close()
+
+        # Set stable to 20 and crash.
```
It's a weird test. It relies on we are not enabling precise checkpoint. Can we instead add a stat and check it to ensure no rts is run?

korteland replied:
I get what you mean about relying on fuzzy checkpoint, but I think this is more robust than a stat were we'd (say) count the number of times we skipped RTS, since we're checking real behaviour rather than some claim about the behaviour. Unless there's some other stat you had in mind?

agorrod reviewed (approved):
Thanks Will. I'm happy with this now - I haven't done a detailed review of code outside the scope of the change, but it appears as though @quchenhao has done that and the included changes lgtm

quchenhao reviewed (approved):
lgtm
